src/core/qgsogrproxytextcodec.cpp

```diff
@@ -0,0 +1,98 @@
+/***************************************************************************
+                             qgsogrproxytextcodec.cpp
+                             -------------
+    begin                : June 2020
+    copyright            : (C) 2020 Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsogrproxytextcodec.h"
+#include "qgslogger.h"
+#include <cpl_string.h>
+#include <mutex>
+
+QgsOgrProxyTextCodec::QgsOgrProxyTextCodec( const QByteArray &name )
+  : mName( name )
+{
+
+}
+
+QString QgsOgrProxyTextCodec::convertToUnicode( const char *chars, int, ConverterState * ) const
+{
+  if ( !chars )
+    return QString();
+
+  char *res = CPLRecode( chars, mName.constData(), CPL_ENC_UTF8 );
+  if ( !res )
+  {
+    QgsDebugMsg( "convertToUnicode failed" );
+    return QString();
+  }
+
+  const QString result = QString::fromUtf8( res );
+  CPLFree( res );
+  return result;
+}
+
+QByteArray QgsOgrProxyTextCodec::convertFromUnicode( const QChar *unicode, int length, ConverterState * ) const
+{
+  if ( !unicode )
+    return QByteArray();
+
+  const QString src = QString( unicode, length );
+  char *res = CPLRecode( src.toLocal8Bit().constData(), CPL_ENC_UTF8, mName.constData() );
+  if ( !res )
+  {
+    QgsDebugMsg( "convertFromUnicode failed" );
+    return QByteArray();
+  }
+
+  const QByteArray result = QByteArray( res );
+  CPLFree( res );
+  return result;
+}
+
+// MY 5 YEAR OLD DAUGHTER WROTE THIS LINE, REMOVE AT YOUR OWN RISK!!!
+// i don't want this to be here
+
+QByteArray QgsOgrProxyTextCodec::name() const
+{
+  return mName;
+}
+
+QList<QByteArray> QgsOgrProxyTextCodec::aliases() const
+{
+  return QList<QByteArray>();
+}
+
+int QgsOgrProxyTextCodec::mibEnum() const
+{
+  // doesn't seem required in this case
+  return 0;
+}
+
+QStringList QgsOgrProxyTextCodec::supportedCodecs()
+{
+  static QStringList codecs;
+  static std::once_flag initialized;
+  std::call_once( initialized, [&]( )
+  {
+    // there's likely others that are supported by GDAL, but we're primarily concerned here
+    // with codecs used by the shapefile driver, and which are no longer supported on the
+    // windows Qt builds (due to removal of ICU support in windows Qt builds)
+    // see https://github.com/qgis/QGIS/issues/36871
+    for ( int i = 437; i <= 950; ++i )
```

Comment:
don't know if that matters in the usage, but not all codes in that range will be valid. On Unix, the list could probably be retrieved from iconv API. On Windows, WideCharToMultiByte() API is used

Reply:
Any clue why Travis is segfaulting with the new test? I can't reproduce locally, and valgrind is happy...

Reply:
I could reproduce it in a Docker container mimicking Travis
The issue is the one pointed at https://doc.qt.io/qt-5/qtextcodec.html#QTextCodec-1

I've confirmed the following change in TestQgsOgrUtils::textCodec() fixes the crash
```
+  // The QTextCodec should always be constructed on the heap. Qt takes ownership and will delete it when the application terminates.
+  QgsOgrProxyTextCodec* codec = new QgsOgrProxyTextCodec( "CP852" );
+  QCOMPARE( codec->toUnicode( codec->fromUnicode( "abcŐ" ) ), QStringLiteral( "abcŐ" ) );
+  QCOMPARE( codec->toUnicode( codec->fromUnicode( "" ) ), QString() );
+  // cppcheck-suppress memleak
```

That also means that qgsvectordataprovider.h must be changed to not free mProxyCodec